### PR TITLE
[VRF]: submit vrf CLI #392 

### PR DESCRIPTION
**Pull request by tylerlinp**
**- What I did**
Submit VRF CLI according to vrf HLD #392

**- How I did it**

**- How to verify it**
Pass the test on nephos lab

**- Previous command output (if the output of a command-line utility has changed)**

```
show ip interfaces
Interface    IPv4 address/mask    Admin/Oper
-----------  -------------------  ------------
Ethernet0    1.1.1.1/24           up/up

show ipv6 interfaces
Interface    IPv6 address/mask                       Admin/Oper
-----------  --------------------------------------  ------------
Ethernet0    fe80::6eec:5aff:fe08:164b%Ethernet0/64  up/up
```

**- New command output (if the output of a command-line utility has changed)**
```
show ip interfaces
Interface    Master    IPv4 address/mask    Admin/Oper
-----------  --------  -------------------  ------------
Ethernet0    Vrf-blue  1.1.1.1/24           up/up

show ipv6 interfaces
Interface    Master    IPv6 address/mask                       Admin/Oper
-----------  --------  --------------------------------------  ------------
Ethernet0    Vrf-blue  fe80::6eec:5aff:fe08:164b%Ethernet0/64  up/up
```
-->

### CODE REVIEW

File: show/main.py [hunk 1]
```diff
@@ -1000,7 +1056,7 @@ def ipv6():
 @ipv6.command()
 def interfaces():
     """Show interfaces IPv6 address"""
-    header = ['Interface', 'IPv6 address/mask', 'Admin/Oper']
+    header = ['Interface', 'Master', 'IPv6 address/mask', 'Admin/Oper']
```
Comment (venkatmahalingam) [inline show/main.py]: Can we have 'VRF' header instead of 'Master'?

File: show/main.py [hunk 2]
```diff
@@ -363,6 +363,47 @@ def cli():
...
+        interface_dict = config_db.get_table(table_name)
+        if interface_dict:
+            for interface in interface_dict.keys():
+                if interface_dict[interface].has_key('vrf_name') and vrf_name == interface_dict[interface]['vrf_name']:
+                    data.append(interface)
+    return data
+
+@cli.command()
+@click.argument('vrf_name', required=False)
+def vrf(vrf_name):
```
Comment (venkatmahalingam) [inline show/main.py]: Please add 'show vrf [vrf-name]' command output as well in the pull request.

File: config/main.py [hunk 3]
```diff
@@ -984,18 +1033,235 @@ def remove(ctx, interface_name, ip_addr):
...
+    pass
+
+@vrf.command('add')
+@click.argument('vrf_name', metavar='<vrf_name>', required=True)
+@click.pass_context
+def add_vrf(ctx, vrf_name):
+    """Add vrf"""
+    config_db = ctx.obj['config_db']
+    if not vrf_name.startswith("Vrf"):
+        ctx.fail("'vrf_name' is not start with Vrf!")
```
Comment (venkatmahalingam) [inline config/main.py]: Do we need to check VRF name length as well and throw an error for invalid len case?

File: config/main.py [hunk 4]
```diff
@@ -984,18 +1033,235 @@ def remove(ctx, interface_name, ip_addr):
...
+    """Bind the interface to VRF"""
+    config_db = ctx.obj["config_db"]
+    if get_interface_naming_mode() == "alias":
+        interface_name = interface_alias_to_name(interface_name)
+        if interface_name is None:
+            ctx.fail("'interface_name' is None!")
+
+    table_name = get_interface_table_name(interface_name)
+    if table_name == "":
+        ctx.fail("'interface_name' is not valid. Valid names [Ethernet/PortChannel/Vlan/Loopback]")
```
Comment (venkatmahalingam) [inline config/main.py]: Do we need to add any support for mgmt intf bind to mgmt VRF (if not present already) as part of this pull request or that should be handled separately?

File: config/main.py [hunk 5]
```diff
@@ -984,18 +1033,235 @@ def remove(ctx, interface_name, ip_addr):
...
+        return
+    interface_dependent = interface_ipaddr_dependent_on_interface(config_db, interface_name)
+    for interface_del in interface_dependent:
+        config_db.set_entry(table_name, interface_del, None)
+    config_db.set_entry(table_name, interface_name, None)
+    # When config_db del entry and then add entry with same key, the DEL will lost.
+    state_db = SonicV2Connector(host='127.0.0.1')
+    state_db.connect(state_db.STATE_DB, False)
+    _hash = '{}{}'.format('INTERFACE_TABLE|', interface_name)
+    while state_db.get(state_db.STATE_DB, _hash, "state") == "ok":
```
Comment (venkatmahalingam) [inline config/main.py]: Why do we need this handling? dont we handle this in VRF-mgrd?

File: config/main.py [hunk 6]
```diff
@@ -984,18 +1033,235 @@ def remove(ctx, interface_name, ip_addr):
...
+            ctx.fail("'interface_name' is None!")
+
+    table_name = get_interface_table_name(interface_name)
+    if table_name == "":
+        ctx.fail("'interface_name' is not valid. Valid names [Ethernet/PortChannel/Vlan/Loopback]")
+    if is_interface_bind_to_vrf(config_db, interface_name) is True and \
+        config_db.get_entry(table_name, interface_name).get('vrf_name') == vrf_name:
+        return
+    interface_dependent = interface_ipaddr_dependent_on_interface(config_db, interface_name)
+    for interface_del in interface_dependent:
```
Comment (venkatmahalingam) [inline config/main.py]: Looks like we're deleting all the IP addresses present on the interface here, please add comment here.

File: config/main.py [hunk 7]
```diff
@@ -984,18 +1033,235 @@ def remove(ctx, interface_name, ip_addr):
...
+        interface_dependent = interface_ipaddr_dependent_on_interface(config_db, interface_name)
+        if len(interface_dependent) == 0 and is_interface_bind_to_vrf(config_db, interface_name) is False:
+            config_db.set_entry(table_name, interface_name, None)
     except ValueError:
         ctx.fail("'ip_addr' is not valid.")
+
+#
+# 'vrf' subgroup ('config interface vrf ...')
+#
+
```
Comment (venkatmahalingam) [inline config/main.py]: Hope we delete the dependent IP, VRF configs upon deleting the portchannel and Vlan...etc?

File: show/main.py (hunk 1 above)
Comment (tylerlinp) [reply show/main.py]: I think 'Master' may be more precise, while device name contains Vrf prefix either.

File: config/main.py (hunk 3 above)
Comment (tylerlinp) [reply config/main.py]: More checks will be better.

File: config/main.py (hunk 4 above)
Comment (tylerlinp) [reply config/main.py]: Now mgmt intf vrf has its own CLI.

File: config/main.py (hunk 5 above)
Comment (tylerlinp) [reply config/main.py]: In intfmgrd. It is a good question. The codes here is to assure DEL op has been handled, or else the new SET op will overwite DEL op and then we have to handle complex intf-move-vrf in intfsorch.

File: config/main.py (hunk 6 above)
Comment (tylerlinp) [reply config/main.py]: OK, I will add later.

File: config/main.py (hunk 7 above)
Comment (tylerlinp) [reply config/main.py]: Deleting dependent configs implicitly can help user to simplify operation, actually it makes CLI implementation more complex.

**Comment by prsunny** [comment]
@tylerlinp , can you rebase to latest code. MGMT VRF [PR](https://github.com/Azure/sonic-utilities/pull/463) is merged, so request to resolve conflicts.

**Comment by tylerlinp** [comment]
retest this please

**Comment by tylerlinp** [comment]
retest this please

**Comment by shine4chen** [comment]
retest this please

**Comment by shine4chen** [comment]
retest this please

**Comment by shine4chen** [comment]
retest this please

**Comment by tylerlinp** [comment]
retest this please

**Comment by tylerlinp** [comment]
retest this please

File: config/main.py [hunk 8]
```diff
@@ -1252,14 +1308,8 @@ def add(ctx, interface_name, ip_addr, gw):
...
-        if interface_name.startswith("Ethernet"):
-            if VLAN_SUB_INTERFACE_SEPARATOR in interface_name:
-                config_db.set_entry("VLAN_SUB_INTERFACE", interface_name, {"admin_status": "up"})
-                config_db.set_entry("VLAN_SUB_INTERFACE", (interface_name, ip_addr), {"NULL": "NULL"})
-            else:
-                config_db.set_entry("INTERFACE", (interface_name, ip_addr), {"NULL": "NULL"})
-                config_db.set_entry("INTERFACE", interface_name, {"NULL": "NULL"})
-        elif interface_name == 'eth0':
+
+        if interface_name == 'eth0':
```
Comment (keboliu) [inline config/main.py]: suggest to don't hardcode here.

Comment (tylerlinp) [reply config/main.py]: It has nothing to do with this PR, it is just merged from master. If to change, a new PR is needed. I think mgmt intf should be handled in another command.

**Comment by prsunny** [comment]
@tylerlinp , can you please resolve conflict?

**Comment by tylerlinp** [comment]
retest this please

**Comment by tylerlinp** [comment]
sonic-swss-build/264 build failed and sonic-swss-build/260 not including PR swss/943, so this PR test failed.

**Comment by prsunny** [comment]
retest this please

**Comment by prsunny** [comment]
retest this please

**Comment by prsunny** [comment]
retest this please

**Comment by prsunny** [comment]
@tylerlinp , can you please check the VS test failures

**Comment by tylerlinp** [comment]
retest this please

**Comment by cony7642** [comment]
How could I use mgmt-vrf configuration? I execute the following  #command(https://github.com/Azure/sonic-buildimage/pull/2585) will get failure information on console.
```
root@sonic:/# config vrf add mgmt
Usage: config vrf add [OPTIONS] <vrf_name>
 
Error: 'vrf_name' is not start with Vrf!
```

**Comment by tylerlinp** [comment]
@cony7642 Oh, that's a issue after this PR merged. I ignored same command defined in 2 PRs.
@prsunny My suggestion is change mgmt vrf commands
```
config vrf add mgmt
config vrf del mgmt
```
to
```
config mgmt-vrf add mgmt
config mgmt-vrf del mgmt
```
and that just like show commands.

**Pull request merged**